### PR TITLE
Allow clients to specify legacy nicknames

### PR DESCRIPTION
**Pull request by tvkanters**
In the [application policies](https://smartdevicelink.com/zh-hans/docs/sdl-server/master/policy-table/application-policies/), each application is assigned a list of nicknames. This field is [used in the SDL Core](https://github.com/smartdevicelink/sdl_core/blob/master/src/components/application_manager/src/commands/mobile/register_app_interface_request.cc#L925) to ensure that an app's name is valid.

When a client connects to an SDL core, app developers can currently only provide one name. In order to allow developers to update their name in the future, they should also be able to provide nicknames that they used in the past to connect to cores with outdated sets of nicknames.

### CODE REVIEW

**Comment by tvkanters** [comment]
@theresalech, this proposal is ready to be reviewed.

File: proposals/NNNN-app-nicknames.md [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+On the [developer profile](https://www.smartdevicelink.com/profile/companies/), developers can specify which nicknames should be valid for their app IDs. However, there is no guarantee that the core's application policies are up-to-date with the latest set of nicknames provided. (E.g., when a car is not configured to have an internet connection.) If an old snapshot of this information is used by the core and a developer updates their app's name, the app would not be able to connect anymore.
+
+For developers, there currently does not seem to be a clean way to work around this issue. It might be possible to try reconnecting to the core with older nicknames in case of a connection failure, but this is not ideal for a number of a reasons:
+* Connectivity will be unnecessarily slowed down.
+* The reconnection attempts may be futile because the core does not give a reason as to _why_ the connection was disallowed.
+* Each app developer would need to create their own logic for this mechanism.
+* If there are issues in a certain client version with creating multiple SDL proxy instances or disposing of old ones (which is not uncommon), the custom solution might not be able work.
```
Comment (theresalech) [inline proposals/NNNN-app-nicknames.md]: Please update the end of this sentence to "might not work."

File: proposals/NNNN-app-nicknames.md [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+
+## Proposed solution
+
+When creating an SDL proxy to instantiate the connection in the client, the developer should be able to provide a list of app names instead of just a single one. This list should be in order of preference, such that preferred names precede less desired legacy names. The core should use this list to determine which name to show for the app. If the first name is in their application policies for the given app ID, it should select that. If not, it should try the second, and so on. If none of the provided names is present, the connection would be disallowed as is currently the case.
```
Comment (theresalech) [inline proposals/NNNN-app-nicknames.md]: Please update the last sentence to "If none of the provided names are present, the connection..."

**Comment by tvkanters** [comment]
Thanks for the rapid feedback! Updated the proposal.

**Pull request merged**